src/js/media/media.js

```diff
@@ -510,7 +511,14 @@ vjs.MediaTechController.withSourceHandlers = function(Tech){
     this.disposeSourceHandler();
     this.off('dispose', this.disposeSourceHandler);
 
-    this.currentSource_ = source;
+    // Set currentSource_ asynchronously to simulate the media element's
+    // asynchronous execution of the `resource selection algorithm`
+    setTimeout(function () {
+      if (source && source.src !== '') {
+        tech.currentSource_ = source;
+      }
+    }, 0);
```

Comment:
If this matches `<video>` behavior better, I think it's the way to go. +1 to `this.setTimeout`.